### PR TITLE
Fix memory corruption at RNG::fill() call 

### LINKED ISSUE #6870: cv::RNG::fill overwrites variables in stack
I tried saving the variables into a xml file, but the saved value has been different from a memory. I felt strange and examined, cv::RNG::fill may overwrite the variables in stack region.

### Please state the information for your system
- OpenCV version: 2.4.9 and 2.4.13, I checked
- Host OS: Linux (Ubuntu 14.04)
- Target OS: host / Linux 3.13.0-86-generic x86_64
- Compiler & CMake: GCC 4.8.4 & CMake 2.8.12.2

### In which part of the OpenCV library you got the issue?
- core

### Expected behavior
All of variables before calling cv::RNG::fill are equal to after. E.g., the following code prints:
(a, b, c, d) = (10, 0.5, -1.3, 104)
(a, b, c, d) = (10, 0.5, -1.3, 104)

### Actual behavior
Some of variables before calling cv::RNG::fill were different from after. E.g., the following code printed:
(a, b, c, d) = (10, 0.5, -1.3, 104)
(a, b, c, d) = (-9, -9, -1.3, 104)

#### Not overwritten when non-double type
I haven't get the same behavior when I changed the type of 'x' to the other type (e.g. cv::Mat_\<uchar\>, cv::Mat_\<cv::Vec2i\>, cv::Mat_\<cv::Point3f\>, etc.).

#### The number of channels may affect the behavior
I have made sure the number of variables in stack has been overwritten.
When the type of 'x' was cv::Mat_\<double\>, no stack variables was overwritten.
When the type of 'x' was cv::Mat_\<cv::Vec2d\> or cv::Mat_\<cv::Point2d\>, only 'a' was overwritten.
When the type of 'x' was cv::Mat_\<cv::Vec3d\> or cv::Mat_\<cv::Point3d\>, 'a' and 'b' were overwritten.

#### Only double-type variables in stack are overwritten
I changed the type of 'a' from double to float in the following code, and then 'a' was not overwritten but 'b' was overwritten.

#### cv::RNG::fill overwrites variables in stack as the value of 'mean'
Overwritten variables in stack had the value of 'mean', which was the 3rd argument of cv::RNG::fill.

#### cv::RNG::fill performs expected behavior when generates uniform distribution
cv::RNG::fill overwrites no variables in stack when I specify cv::RNG::UNIFORM to the 2nd argument of cv::RNG::fill.

### Code example to reproduce the issue / Steps to reproduce the issue
```
#include <iostream>
#include <opencv2/core.hpp>

int main() {
    double a = 10;
    double b = 0.5;
    cv::Mat_<cv::Point2d> x(2, 3);
    double c = -1.3;
    double d = 104;

    std::cerr << "(a, b, c, d) = " << a << ", " << b << ", " << c << ", " << d << ")" << std::endl;
    cv::theRNG().fill(x, cv::RNG::NORMAL, -9, 10);
    std::cerr << "(a, b, c, d) = " << a << ", " << b << ", " << c << ", " << d << ")" << std::endl;

    return 0;
}
```

### PR DESCRIPTION
**Pull request by savuor**
resolves  #6870

### What does this PR change?
A check added for the case when normal distribution parameters have less channels than the matrix to be filled.

### CODE REVIEW

**Comment by apavlenko** [comment]
:+1:

**Comment by apavlenko** [comment]
@savuor , are you going to port the fix to 2.4?

**Comment by ilya-lavrenov** [comment]
I think you should add the provided reproducer as a regression test.

**Comment by apavlenko** [comment]
agree with @ilya-lavrenov !
@savuor , please add the test to this PR

File: modules/core/test/test_rand.cpp [hunk 1]
```diff
@@ -365,3 +365,21 @@ TEST(Core_RNG_MT19937, regression)
...
     }
 }
+
+
+TEST(Core_Rand, Regression_Stack_Corruption)
+{
+    int bufsz = 128; //enough for 14 doubles
+    AutoBuffer<uchar> buffer(bufsz);
+    size_t offset = 0;
+    cv::Mat_<cv::Point2d> x(2, 3, (cv::Point2d*)(buffer+offset)); offset += sizeof(cv::Mat_<cv::Point2d>);
```
Comment (alalek) [inline modules/core/test/test_rand.cpp]: sizeof(cv::Mat_<cv::Point2d>)
=>
x.total()*x.elemsize()

**Comment by savuor** [comment]
@apavlenko done
backport added: #6920 
As it turned out, RNG::fill() corrupts not just stack but any memory containing its arguments

File: modules/core/test/test_rand.cpp (hunk 1 above)
Comment (savuor) [inline modules/core/test/test_rand.cpp]: @alalek fixed

File: modules/core/test/test_hal_core.cpp [hunk 2]
```diff
@@ -179,8 +179,8 @@ TEST(Core_HAL, mat_decomp)
             t = (double)getTickCount() - t;
             min_ocv_t = std::min(min_ocv_t, t);
         }
-        //std::cout << "x: " << Mat(x.t()) << std::endl;
-        //std::cout << "x0: " << Mat(x0.t()) << std::endl;
+        std::cout << "x: " << Mat(x.t()) << std::endl;
```
Comment (mshabunin) [inline modules/core/test/test_hal_core.cpp]: Please remove or comment these lines.

**Pull request merged**